ufl/formoperators.py

```diff
@@ -304,9 +304,21 @@ def derivative(form, coefficient, argument=None, coefficient_derivatives=None):
     ``Coefficient`` instances to their derivatives w.r.t. *coefficient*.
     """
     if isinstance(form, FormSum):
-        # Distribute derivative over FormSum components
-        return FormSum(*[(derivative(component, coefficient, argument, coefficient_derivatives), 1)
-                         for component in form.components()])
+        # Distribute derivative over FormSum components and weights that depend on coefficient
```

Comment:
This won't do anything if you don't also make sure the DAG visitor is aware that it needs to visit the weights during differentiation, see https://github.com/firedrakeproject/ufl/blob/cf66c4b371604e53854459dd8bcd7c3b5f8ee5cf/ufl/algorithms/map_integrands.py#L39. There is a couple of things to change to support weights that are non constant, see my comment in #44.